File: packages/node/src/routing/handlers/command.ts
import type { Update } from "../../kernel/update.js";
import { CallbackContext } from "../../kernel/context.js";
import { BaseFilter } from "../../filters/base.js";
import { BaseHandler, HandlerCallback } from "./base.js";

/**
 * Handler for Telegram bot commands (e.g. `/start`, `/help`).
 *
 * @typeParam C - Type of the callback context.
 * @typeParam R - Return value type.
 * @example
 * ```ts
 * const startHandler = new CommandHandler("start", async (update, context) => {
 *   await context.bot.sendMessage({ chat_id: update.effective_chat!.id, text: "Hello!" });
 * });
 * ```
 */
export class CommandHandler<
  C extends CallbackContext = CallbackContext,
  R = unknown,
> extends BaseHandler<C, R> {
  /**
   * Normalized set of command names that trigger this handler (without `/` or `@botname`).
   */
  public readonly commands: Set<string>;
  /**
   * Optional secondary filter to apply before triggering.
   */
  public readonly filters?: BaseFilter;

  /**
   * Creates a new {@link CommandHandler}.
   *
   * @param command - A single command string (e.g. `"start"`) or an array of aliases.
   * @param callback - Function invoked when the command matches.
   * @param filters - Optional additional filter criteria.
   * @throws When command is empty or contains only whitespace.
   */
  constructor(command: string | string[], callback: HandlerCallback<C, R>, filters?: BaseFilter) {
    super(callback);
    const commandList = Array.isArray(command) ? command : [command];
    if (commandList.length === 0 || commandList.some((c) => !c || c.trim() === "")) {
      throw new Error("CommandHandler requires at least one non-empty command string.");
    }

    this.commands = new Set(commandList.map((c) => c.toLowerCase()));
    this.filters = filters;
  }

  /**
   * Checks whether the update contains a matching bot command.
   *
   * @param update - The update to test.
   * @returns `true` if command matches, `false` otherwise.
   */
  async checkUpdate(update: Update): Promise<boolean> {
    const msg = update.effective_message;
    if (!msg || !msg.text) return false;
    if (!msg.entities || msg.entities.length === 0) return false;
    const firstEntity = msg.entities[0];
    if (!firstEntity || firstEntity.type !== "bot_command" || firstEntity.offset !== 0) {
      return false;
    }

    const commandText = msg.text.slice(1, firstEntity.length);
    const [commandName] = commandText.split("@");
    if (!commandName || !this.commands.has(commandName.toLowerCase())) {
      return false;
    }

    if (this.filters) {
      const match = await this.filters.checkUpdate(update);
      if (!match) return false;
    }

    return true;
  }

  /**
   * Parses positional command arguments into `context.args` and executes callback.
   *
   * @param update - The incoming update.
   * @param context - Callback context instance.
   * @returns Result from callback execution.
   */
  override async handleUpdate(update: Update, context: C): Promise<R> {
    const msg = update.effective_message;
    if (msg?.text) {
      const parts = msg.text.trim().split(/\s+/);
      context.args = parts.slice(1);
    }

    return super.handleUpdate(update, context);
  }
}

File: packages/python/src/telebot_py/routing/handlers/command.py
"""Handler for Telegram bot commands such as ``/start``."""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any

from telebot_py.filters.base import MessageFilter
from telebot_py.routing.handlers.base import BaseHandler, HandlerCallback
from telebot_py.types import Update


class CommandHandler(BaseHandler):
    """Handler for bot commands like ``/start``.

    Command names are compared against the message text with the leading
    ``/`` and any ``@BotName`` mention suffix stripped, case-insensitively.
    The words following the command are exposed on the callback context as
    ``context.args``.

    Attributes:
        commands: Normalized (lower-cased, slash-less) command names.
        filters: Optional message filter gating the match after the command
            name matched.
    """

    def __init__(
        self,
        command: str | Sequence[str],
        callback: HandlerCallback,
        *,
        filters: MessageFilter | None = None,
    ) -> None:
        """Initialize the handler.

        Args:
            command: One command name (with or without leading ``/``) or a
                sequence of aliases.
            callback: Callable invoked when a matching command arrives.
            filters: Optional additional message filter.

        Raises:
            ValueError: If ``command`` is empty or contains a blank entry
                (FR-014 fail-fast on programmer errors).
        """
        super().__init__(callback)
        names = [command] if isinstance(command, str) else list(command)
        if not names:
            msg = "CommandHandler requires at least one command."
            raise ValueError(msg)
        normalized: list[str] = []
        for name in names:
            cleaned = name.strip().removeprefix("/")
            if not cleaned:
                msg = "CommandHandler requires non-empty command strings."
                raise ValueError(msg)
            normalized.append(cleaned.lower())
        self.commands: frozenset[str] = frozenset(normalized)
        self.filters = filters

    def check_update(self, update: object) -> list[str] | None:
        """Return the command's argument list when the update matches.

        Args:
            update: The incoming update.

        Returns:
            The argument words following the command (possibly empty) on a
            match, ``None`` otherwise.
        """
        if not isinstance(update, Update):
            return None
        message = update.effective_message
        if message is None or not message.text or not message.text.startswith("/"):
            return None
        parts = message.text.split()
        command = parts[0][1:]
        if "@" in command:
            command = command.split("@", 1)[0]
        if not command or command.lower() not in self.commands:
            return None
        if self.filters is not None and not self.filters(message):
            return None
        return parts[1:]

    def collect_additional_context(self, context: Any, update: Update, check_result: Any) -> None:
        """Store the parsed command arguments on ``context.args``.

        Args:
            context: The callback context being populated.
            update: The update being processed.
            check_result: The argument list :meth:`check_update` returned.
        """
        if check_result is not None:
            context.args = check_result
